core/src/main/java/org/commonjava/maven/ext/core/impl/PreparseGroovyManipulator.java

```diff
@@ -0,0 +1,134 @@
+
+import javax.inject.Inject;
+import javax.inject.Named;
+import javax.inject.Singleton;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.util.Collections;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * {@link Manipulator} implementation that can resolve a remote groovy file and execute it on executionRoot. Configuration
+ * is stored in a {@link GroovyState} instance, which is in turn stored in the {@link ManipulationSession}.
+ */
+@Named("groovy-injection-preparse")
+@Singleton
+public class PreparseGroovyManipulator
+                extends BaseGroovyManipulator
+    //implements Manipulator
+{
+    @Inject
+    public PreparseGroovyManipulator( ModelIO modelIO, FileIO fileIO, PomIO pomIO )
+    {
+        super( modelIO, fileIO, pomIO );
+    }
+
+    /**
+     * Initialize the {@link GroovyState} state holder in the {@link ManipulationSession}. This state holder detects
+     * version-change configuration from the Maven user properties (-D properties from the CLI) and makes it available
+     * for later.
+     */
+    //@Override
+    public void init( final ManipulationSession session ) throws ManipulationException
+    {
+        final State state = new GroovyState( session.getUserProperties() );
+        this.session = session;
+        session.setState( state );
+
+        try
+        {
+            final GalleyInfrastructure infra = new GalleyInfrastructure( session, new DefaultMirrorSelector() );
+            infra.init( null, null, Files.createTempDirectory( "pme-" ).resolve( "cache-dir" ).toFile() );
+            final GalleyAPIWrapper wrapper = new GalleyAPIWrapper( infra );
+            modelIO = new ModelIO( wrapper );
+            fileIO = new FileIO( infra );
+        }
+        catch ( IOException e )
+        {
+            throw new ManipulationException( e.getMessage(), e );
+        }
+    }
+
+    /**
+     * Apply the groovy script changes.
+     */
+    //@Override
+    public Set<Project> applyChanges( final List<Project> projects )
+        throws ManipulationException
+    {
+        final GroovyState state = session.getState( GroovyState.class );
+
+        if ( !session.isEnabled() || !state.isEnabled() )
+        {
+            logger.debug("{}: Nothing to do!", getClass().getSimpleName() );
+            return Collections.emptySet();
+        }
+
+        final List<File> groovyScripts = parseGroovyScripts( state.getGroovyScripts() );
+
+        for ( final File groovyScript : groovyScripts )
+        {
+            final GroovyShell shell = new GroovyShell();
+
+            try
+            {
+                final Script script = shell.parse(groovyScript);
+                final InvocationPoint invocationPoint = script.getClass().getAnnotation(InvocationPoint.class);
+                final InvocationStage stage = invocationPoint != null ? invocationPoint.invocationPoint() : null;
+
+                if ( InvocationStage.PREPARSE == stage )
```

Comment:
Why are lines 106->114 needed as can't `applyGroovyScript` be used? Possibly the info logger in that apply function needs to be a bit cleverer though.

Reply:
It was currently to block execution on `stage != PREPARSE` since the `if` inside `applyGroovyScript` currently runs with `BOTH`, but as I explained elsewhere, running `PREPARSE` in that case is going to break existing tests.

Reply:
Can you point me to an example ? Can't the tests only run with first/last phases?

Reply:
I've removed the `if` from this class, so now it should be a question of whether the `if` in `BaseGroovyManipulator` looks correct.